### PR TITLE
feat: add new Location Provider component

### PR DESCRIPTION
**Pull request by pmareke**
EX-4639

Add new `LocationProvider` component.

# Pull request template
 
This PR adds a new component which receives a prop called `location` and inject it in its child.

## Motivation and context

With the new **Tagging** feature coming, we need to be able to provide custom locations in our components, this component make it possible.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Change requires a documentation update

## What is the destination branch of this PR?

- [x] `Main`
- [ ] Other. Specify:

## How has this been tested?

https://www.loom.com/share/415361876dd3417fa88a4b05ba139472

Tests performed according to [testing guidelines](./contributing/tests.md): 

## Checklist:

- [x] My code follows the **[style guidelines](./CONTRIBUTING.md#style-guides)** of this project.
- [x] I have performed a **self-review** on my own code.
- [x] I have **commented** my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the **[documentation](./CONTRIBUTING.md#documentation-style-guide)**.
- [x] My changes generate **no new warnings**.
- [x] I have added **tests** that prove my fix is effective or that my feature works.
- [x] New and existing **unit tests pass locally** with my changes.

### CODE REVIEW

File: packages/x-components/src/components/__tests__/location-injector.spec.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    },
+    {
+      propsData: {
+        location
+      }
+    }
+  );
+
+  return {
+    child: wrapper.find(getDataTestSelector('inject') as VueClass<Child>)
```
Comment (javieri-empathy) [inline packages/x-components/src/components/__tests__/location-injector.spec.ts:45]: There is something wrong with vue-test-utils types that I don't get. You should be able to use 
` wrapper.findComponent(Child)`, but type inference does not work and forces you to write the generic instead ` wrapper.findComponent<Child>(Child)`. IMO this is easier than writing a data-test selector and cast it.

Comment (javieri-empathy) [inline packages/x-components/src/components/__tests__/location-injector.spec.ts:45]: There is something wrong with type inference and vue-test-utils. You should be able to write
```
    child: wrapper.findComponent(Child)
```

but type inference does not work and makes you explicitly type the generic:

```suggestion
    child: child: wrapper.findComponent<Child>(Child)
```

IMO this is easier than using a data-test selector and cast it. **Your way is fine**, I'm just saying this as free info.

File: packages/x-components/src/components/location-injector.vue [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    /**
+     * {@link FeatureLocation | Location} To inject.
+     *
+     * @public
+     */
+    @Prop({ required: true })
+    @Provide('location')
+    protected location!: FeatureLocation;
+  }
+</script>
```
Comment (javieri-empathy) [inline packages/x-components/src/components/location-injector.vue:23]: Where are the docs of this component?

File: packages/x-components/src/components/__tests__/location-injector.spec.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+@Component({
+  template: `<div data-test="inject"/>`
+})
+class Child extends Vue {
+  @Inject('location')
+  public location!: FeatureLocation;
+}
+
+/**
+ * Renders the `LocationInjector` component, exposing a basic API for testing.
```
Comment (javieri-empathy) [inline packages/x-components/src/components/__tests__/location-injector.spec.ts:17]: Whenever it is possible, try to link documentation parts like the name of the components so it is easier to refactor if necessary:

```suggestion
 * Renders the {@link LocationInjector} component, exposing a basic API for testing.
```

File: packages/x-components/src/components/location-injector.vue [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  /**
+   * Location Injector component.
+   * This component injects the location with value passed as prop.
+   *
+   * @public
+   */
+  @Component
+  export default class LocationInjector extends NoElement {
+    /**
+     * {@link FeatureLocation | Location} To inject.
```
Comment (javieri-empathy) [inline packages/x-components/src/components/location-injector.vue:15]: Do not rename the type if not necessary. Also if you add a *The* you don't have to wrongly capitalise the *To*.
```suggestion
      * The {@link FeatureLocation} to inject.
```

File: packages/x-components/src/components/__tests__/location-injector.spec.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    const { child } = renderLocationInjector({
+      location: 'external'
+    });
+
+    expect(child.vm.location).toBe('external');
+  });
+});
+
+/**
+ * Options to configure how the progress bar component should be rendered.
```
Comment (javieri-empathy) [inline packages/x-components/src/components/__tests__/location-injector.spec.ts:60]: A progress bar? Isn't this the test for the location injector? 🤔

File: packages/x-components/src/components/__tests__/location-injector.spec.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ */
+interface RenderLocationInjectorOptions {
+  /** The location to inject. */
+  location: FeatureLocation;
+}
+
+/**
+ * Options to configure how the location injector component should be rendered.
```
Comment (javieri-empathy) [inline packages/x-components/src/components/__tests__/location-injector.spec.ts:68]: This is not the options to configure nothing, this is the helper functions for testing the component.

**Review by javieri-empathy — CHANGES_REQUESTED** [review]

File: packages/x-components/src/components/__tests__/location-injector.spec.ts (hunk 1 above)
Comment (pmareke) [reply packages/x-components/src/components/__tests__/location-injector.spec.ts:45]: Nice one, thanks.

File: packages/x-components/src/components/__tests__/location-injector.spec.ts (hunk 6 above)
Comment (pmareke) [reply packages/x-components/src/components/__tests__/location-injector.spec.ts:68]: My fault, taking a look to the rest of the components we usually don't comment these kind of interfaces.

File: packages/x-components/src/components/location-injector.vue (hunk 2 above)
Comment (pmareke) [reply packages/x-components/src/components/location-injector.vue:26]: Sorry.

File: packages/x-components/src/components/location-injector.vue [hunk 7]
```diff
@@ -0,0 +1,62 @@
+<script lang="ts">
+  import { Prop, Provide, Component } from 'vue-property-decorator';
+  import { NoElement } from '../components/no-element';
+  import { FeatureLocation } from '../types';
+  import BaseEventButton from './base-event-button.vue';
```
Comment (LuisMartinez15) [inline packages/x-components/src/components/location-injector.vue:5]: Remove unused `import`

```suggestion
```

File: packages/x-components/src/components/location-injector.vue [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * Location Injector component.
+   * This component injects the location with value passed as prop.
+   *
+   * @public
+   */
+  @Component({
+    components: { BaseEventButton }
+  })
```
Comment (LuisMartinez15) [inline packages/x-components/src/components/location-injector.vue:15]: Remove unused registered component

```suggestion
  @Component
```

File: packages/x-components/src/components/location-injector.vue [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  export default class LocationInjector extends NoElement {
+    /**
+     * The {@link FeatureLocation} to inject.
+     *
+     * @public
+     */
+    @Prop({ required: true })
+    @Provide('location')
```
Comment (LuisMartinez15) [inline packages/x-components/src/components/location-injector.vue:23]: `Vue`'s `Provide` uses by default the same key as the variable name ([docs](https://github.com/kaorun343/vue-property-decorator#Provide)). So, in this case, you don't need to specify it again.

```suggestion
    @Provide()
```

File: packages/x-components/src/components/location-injector.vue [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+   * Location Injector component.
+   * This component injects the location with value passed as prop.
+   *
+   * @public
+   */
+  @Component({
+    components: { BaseEventButton }
+  })
+  export default class LocationInjector extends NoElement {
```
Comment (LuisMartinez15) [inline packages/x-components/src/components/location-injector.vue:16]: Can you rename it to `LocationProvider`? It is providing a location, not injecting it 😅

File: packages/x-components/src/components/location-injector.vue [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * The {@link FeatureLocation} to inject.
```
Comment (LuisMartinez15) [inline packages/x-components/src/components/location-injector.vue:18]:
```suggestion
     * The {@link FeatureLocation} to provide.
```

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
Good job mate 💪🏿 💪🏿

**Review by LuisMartinez15 — DISMISSED** [review]
LVIPGTM

File: packages/x-components/src/components/location-provider.vue [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+      };
+    }
+  };
+</script>
+```
+
+## Events
+
+This component doesn't emit events.
+</docs>
```
Comment (javieri-empathy) [inline packages/x-components/src/components/location-provider.vue:59]: A few advices that will improve the doc IMO: 

1. Lets try to follow the same structure for new components than the ones that have been validated by eDocs, like the one in `SearchInput`.
2. `BaseEventButton` component does not expect an array of events, but an object. The `UserAcceptedAQuery` event should always have a payload. In any case, I would use a real example.
3. This examples are supposedly going to be rendered, as this component is not very visual, and it all about the metadata in the events, let's try to leave it crystal clear why you need this.

This is my proposal:

```suggestion
## Events

This component doesn't emit events. However, events emitted by child components, will include the
`location` prop value in its metadata.

## See it in action

In this example, the `LocationProvider` component receives the `location` prop with the
`"predictive_layer"` value. This value will then be combined with the `NextQueries` feature name,
`next_query` to create a `QueryOrigin` string, which is used for tracking the performance of the
different tools that generate the query.

```vue
<template>
  <LocationProvider location="predictive_layer">
    <NextQueries :events="events" />
  </LocationProvider>
</template>

<script>
  import { LocationProvider } from '@empathyco/x-components';
  import { NextQueries } from '@empathyco/x-components/next-queries';

  export default {
    name: 'LocationProviderDemo',
    components: {
      NextQueries,
      LocationProvider
    }
  };
</script>
```
```

Comment (pmareke) [reply packages/x-components/src/components/location-provider.vue:59]: You're completely right, I'll remove the events from the NextQueries element

`<NextQueries :events="events" />`

as the rest s perfect.

**Review by javieri-empathy — APPROVED** [review]

**Review by LuisMartinez15 — APPROVED** [review]

**Pull request merged**